File: docs/examples/partner-client.mjs
#!/usr/bin/env node
/**
 * Minimal Node.js partner client (Task 02).
 * Run:
 *   JOHNNYONE_BASE_URL=https://johnnyone-dev.ethan-353.workers.dev \
 *   JOHNNYONE_EMAIL=... JOHNNYONE_PW=... JOHNNYONE_TENANT=... \
 *   node docs/examples/partner-client.mjs
 *
 * Or supply a pre-issued token:
 *   JOHNNYONE_BASE_URL=... JOHNNYONE_TOKEN=... \
 *   JOHNNYONE_PROVIDER=ollama JOHNNYONE_WORKDIR=/tmp \
 *   node docs/examples/partner-client.mjs
 *
 * Requires: npm i ws   (uses global fetch on Node 18+)
 *
 * Flow: login (or token) → createAiSession → WSS (Bearer) → subscribe → send command → read updated screen.
 * Errors: use WSS upgrade HTTP status (401/503 via error/unexpected-response), terminal_command_ack.error (e.g. 'forbidden_session'), GraphQL errors[].message.
 * Envelopes match chat-relay-do.ts RelayEnvelope exactly.
 */

import WebSocket from 'ws';

const BASE = process.env.JOHNNYONE_BASE_URL || 'https://johnnyone-dev.ethan-353.workers.dev';
const TOKEN = process.env.JOHNNYONE_TOKEN;
const EMAIL = process.env.JOHNNYONE_EMAIL || 'partner@example.com';
const PW = process.env.JOHNNYONE_PW;
const TENANT = process.env.JOHNNYONE_TENANT || 'your-tenant';
const PROVIDER = process.env.JOHNNYONE_PROVIDER || 'ollama';
const WORKDIR = process.env.JOHNNYONE_WORKDIR || '/tmp/partner-demo';

const gql = async (query, variables, token, tenantId) => {
  const res = await fetch(`${BASE}/graphql`, {
    method: 'POST',
    headers: {
      'content-type': 'application/json',
      ...(tenantId ? { 'x-tenant-id': tenantId } : {}),
      ...(token ? { authorization: `Bearer ${token}` } : {})
    },
    body: JSON.stringify({ query, variables })
  });
  if (!res.ok) throw new Error(`GraphQL ${res.status}`);
  const j = await res.json();
  if (j.errors?.length) {
    const msg = j.errors[0].message;
    if (msg.includes('Desktop not connected') || msg.includes('503')) {
      throw new Error('503 no online desktop node');
    }
    throw new Error(msg);
  }
  return j;
};

(async () => {
  let token = TOKEN;
  if (!token) {
    if (!PW) throw new Error('Provide JOHNNYONE_TOKEN or JOHNNYONE_PW');
    const { data: { login } } = await gql(
      'mutation($i:LoginInput!){login(input:$i){accessToken refreshToken}}',
      { i: { email: EMAIL, password: PW, tenantId: TENANT } },
      null,
      TENANT
    );
    token = login.accessToken;
    console.error('Logged in, got accessToken');
  }

  // create session
  const { data: { createAiSession: s } } = await gql(
    'mutation($i:CreateAiSessionInput!){createAiSession(input:$i){id}}',
    { i: { provider: PROVIDER, workingDirectory: WORKDIR } },
    token
  );
  const sessionId = s.id;
  console.error('Session created:', sessionId);

  // WSS
  const wsUrl = BASE.replace(/^https/, 'wss') + '/api/relay/ws';
  const ws = new WebSocket(wsUrl, { headers: { authorization: `Bearer ${token}` } });

  const timeout = setTimeout(() => { console.error('timeout'); ws.close(); process.exit(1); }, 15000);

  let firstScreen = null;
  ws.on('open', () => {
    ws.send(JSON.stringify({ type: 'terminal_visual_subscribe', sessionId }));
  });

  ws.on('message', (buf) => {
    const msg = JSON.parse(buf.toString());
    if (msg.type === 'terminal_screen') {
      console.log('SCREEN:', JSON.stringify(msg.data));
      if (!firstScreen) {
        firstScreen = msg.data;
        console.error('received first screen');
        // now send command to prove updated screen
        ws.send(JSON.stringify({ type: 'terminal_command', sessionId, data: 'echo partner-demo\r' }));
      } else {
        // updated screen after command
        clearTimeout(timeout);
        ws.close();
        console.error('received updated screen after command');
        if (JSON.stringify(msg.data) !== JSON.stringify(firstScreen)) {
          console.error('SUCCESS: updated screen differs from first (input round-trip observed)');
        } else {
          console.error('SUCCESS: received screen after command (input accepted)');
        }
        process.exit(0);
      }
    } else if (msg.type === 'terminal_command_ack') {
      if (msg.error) {
        if (msg.error === 'forbidden_session') {
          console.error('ERROR: forbidden_session (session not owned by this tenant/user)');
        } else {
          console.error('ERROR: ' + msg.error);
        }
        process.exit(1);
      }
      console.error('command ack');
    }
  });

  ws.on('error', (e) => {
    const msg = e.message || String(e);
    if (msg.includes('401') || msg.includes('Unauthorized')) {
      console.error('ERROR: 401 unauthorized on WSS upgrade (invalid/missing token)');
    } else if (msg.includes('503') || msg.includes('Service Unavailable') || msg.includes('no online') || msg.includes('Desktop not connected')) {
      console.error('ERROR: 503 no online desktop node');
    } else {
      console.error('WS error:', msg);
    }
    process.exit(1);
  });
  ws.on('close', (c, r) => {
    if (c === 1008 || c === 4001) console.error('ERROR: 401 unauthorized on WSS upgrade (invalid/missing token)');
    else if (c === 1013 || String(r).includes('503')) console.error('ERROR: 503 no online desktop node');
    else if (c !== 1000) console.error('WS closed', c, r.toString());
  });
  ws.on('unexpected-response', (req, res) => {
    const status = res.statusCode;
    if (status === 401) console.error('ERROR: 401 unauthorized on WSS upgrade (invalid/missing token)');
    else if (status === 503) console.error('ERROR: 503 no online desktop node');
    else console.error('WS unexpected response:', status);
    process.exit(1);
  });
})().catch(e => {
  if (e.message && e.message.includes('401')) console.error('ERROR: 401 on GraphQL (auth failed)');
  else if (e.message && (e.message.includes('503') || e.message.includes('no online desktop node'))) console.error('ERROR: 503 no online desktop node');
  else console.error('FAIL:', e.message);
  process.exit(1);
});

File: docs/examples/partner-client.py
#!/usr/bin/env python3
"""
Minimal Python partner client (Task 03).
Run:
  python -m pip install httpx websockets
  JOHNNYONE_BASE_URL=https://johnnyone-dev.ethan-353.workers.dev \
  JOHNNYONE_EMAIL=... JOHNNYONE_PW=... JOHNNYONE_TENANT=... \
  python docs/examples/partner-client.py

Or with token:
  JOHNNYONE_BASE_URL=... JOHNNYONE_TOKEN=... \
  JOHNNYONE_PROVIDER=ollama JOHNNYONE_WORKDIR=/tmp \
  python docs/examples/partner-client.py

Uses httpx + websockets (not pure stdlib). Full flow: login or token → create → WSS (header or ?token) → subscribe → command → updated screen.
Envelopes exactly as in chat-relay-do.ts.
Errors: GraphQL errors[].message or HTTP status; WSS upgrade 401/503; terminal_command_ack.error:'forbidden_session'.
Note: uses additional_headers= (websockets >=12); ?token= fallback documented for older header-less clients.
"""

import asyncio
import json
import os
import sys

try:
    import httpx
    import websockets
except ImportError:
    print("pip install httpx websockets", file=sys.stderr)
    sys.exit(1)

BASE = os.getenv("JOHNNYONE_BASE_URL", "https://johnnyone-dev.ethan-353.workers.dev")
TOKEN = os.getenv("JOHNNYONE_TOKEN")
EMAIL = os.getenv("JOHNNYONE_EMAIL", "partner@example.com")
PW = os.getenv("JOHNNYONE_PW")
TENANT = os.getenv("JOHNNYONE_TENANT", "your-tenant")
PROVIDER = os.getenv("JOHNNYONE_PROVIDER", "ollama")
WORKDIR = os.getenv("JOHNNYONE_WORKDIR", "/tmp/partner-demo")

async def main():
    token = TOKEN
    async with httpx.AsyncClient() as client:
        if not token:
            if not PW:
                raise SystemExit("Provide JOHNNYONE_TOKEN or JOHNNYONE_PW")
            r = await client.post(
                f"{BASE}/graphql",
                headers={"content-type": "application/json", "x-tenant-id": TENANT},
                json={
                    "query": "mutation($i:LoginInput!){login(input:$i){accessToken refreshToken}}",
                    "variables": {"i": {"email": EMAIL, "password": PW, "tenantId": TENANT}},
                },
            )
            resp = r.json()
            if resp.get("errors"):
                err_msg = resp["errors"][0].get("message", "")
                if "401" in err_msg or "unauthorized" in err_msg.lower():
                    print("ERROR: 401 unauthorized on GraphQL (invalid/missing token)", file=sys.stderr)
                else:
                    print("ERROR: " + err_msg, file=sys.stderr)
                sys.exit(1)
            r.raise_for_status()
            data = resp.get("data") or {}
            token = data.get("login", {}).get("accessToken")
            if not token:
                print("ERROR: 401 unauthorized on GraphQL", file=sys.stderr)
                sys.exit(1)
            print("logged in", file=sys.stderr)

        # create
        r = await client.post(
            f"{BASE}/graphql",
            headers={"authorization": f"Bearer {token}"},
            json={
                "query": "mutation($i:CreateAiSessionInput!){createAiSession(input:$i){id}}",
                "variables": {"i": {"provider": PROVIDER, "workingDirectory": WORKDIR}},
            },
        )
        resp = r.json()
        if resp.get("errors"):
            err_msg = resp["errors"][0].get("message", "")
            if "401" in err_msg or "unauthorized" in err_msg.lower():
                print("ERROR: 401 unauthorized on GraphQL (invalid/missing token)", file=sys.stderr)
            elif "503" in err_msg or "desktop not connected" in err_msg.lower() or "no online" in err_msg.lower():
                print("ERROR: 503 no online desktop node", file=sys.stderr)
            elif "forbidden" in err_msg.lower() or "forbidden_session" in err_msg:
                print("ERROR: forbidden_session (session not owned by this tenant/user)", file=sys.stderr)
            else:
                print("ERROR: " + err_msg, file=sys.stderr)
            sys.exit(1)
        r.raise_for_status()
        data = resp.get("data") or {}
        session_id = data.get("createAiSession", {}).get("id")
        if not session_id:
            print("ERROR: 503 no online desktop node (create failed)", file=sys.stderr)
            sys.exit(1)
        print("session", session_id, file=sys.stderr)

        # WSS - prefer header (use additional_headers for websockets >=12; extra_headers was used in <12 and removed in >=14)
        ws_url = BASE.replace("https://", "wss://") + "/api/relay/ws"
        headers = {"Authorization": f"Bearer {token}"}

        async with websockets.connect(ws_url, additional_headers=headers) as ws:
            await ws.send(json.dumps({"type": "terminal_visual_subscribe", "sessionId": session_id}))

            first_screen = None
            async for msg in ws:
                m = json.loads(msg)
                if m.get("type") == "terminal_command_ack" and m.get("error"):
                    err = m.get("error")
                    if err == "forbidden_session":
                        print("ERROR: forbidden_session (session not owned by this tenant/user)", file=sys.stderr)
                    else:
                        print("ERROR: " + err, file=sys.stderr)
                    return
                if m.get("type") == "terminal_screen":
                    print("SCREEN:", json.dumps(m.get("data")))
                    if first_screen is None:
                        first_screen = m.get("data")
                        print("received first screen", file=sys.stderr)
                        # now send command to prove updated screen
                        await ws.send(json.dumps({"type": "terminal_command", "sessionId": session_id, "data": "echo python-demo\r"}))
                    else:
                        print("received updated screen after command", file=sys.stderr)
                        if m.get("data") != first_screen:
                            print("SUCCESS: updated screen differs from first (input round-trip observed)", file=sys.stderr)
                        else:
                            print("SUCCESS: received screen after command (input accepted)", file=sys.stderr)
                        return
            raise SystemExit("no screen received")

if __name__ == "__main__":
    try:
        asyncio.run(main())
    except Exception as e:
        msg = str(e)
        if '401' in msg or 'unauthorized' in msg.lower():
            print("ERROR: 401 unauthorized on WSS/GraphQL (invalid/missing token)", file=sys.stderr)
        elif '503' in msg or 'no online' in msg.lower():
            print("ERROR: 503 no online desktop node", file=sys.stderr)
        elif 'forbidden_session' in msg:
            print("ERROR: forbidden_session (session not owned by this tenant/user)", file=sys.stderr)
        else:
            print("FAIL:", e, file=sys.stderr)
        sys.exit(1)
